### PR TITLE
[DistillationTrainer refactor] Document the num_items_in_batch denominator bug

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 0.3 — stacked on #6451. Last PR of the safety-net phase. Part of #6449.*

The divergence loss is reduced as `sum / num_items_in_batch`, so that count is the loss denominator. It is currently wrong.

`transformers.Trainer` computes it from the raw dataloader batches, before `_prepare_inputs` runs. Two things follow:

- `_RepeatBatchDataLoader` yields the same generation batch once per accumulation step, so the count is `gradient_accumulation_steps` times too large.
- It is derived from the **dataset** labels, even on steps whose completions are replaced by generated ones.

The effect is a silently down-scaled loss and gradient.

Adds an xfail test that records what the trainer is actually handed during `train()` and compares it against the tokens the loss is summed over.

No production code touched — this documents the bug rather than fixing it, so the eventual fix lands as a visible, intentional change.

### Note

The existing `test_loss_normalizes_by_num_items_in_batch` passes `num_items_in_batch` explicitly, so it exercises the *reduction* rather than the *value*, and cannot catch this.

Un-xfail in the PR that moves the generation buffer to the GRPO-style `_prepare_inputs` (plan step 5.6), where the denominator becomes the global completion-token count.

### Risk

Low — tests only. Marked non-strict xfail, so it reports XPASS rather than failing if the behaviour turns out to differ from the analysis.

---

> [!NOTE]
> **Low Risk**
> Tests only; non-strict xfail documents behavior without changing training logic.
> 
> **Overview**
> Adds **`test_num_items_in_batch_counts_the_tokens_trained_on`**, marked **`@pytest.mark.xfail`**, to lock in expected behavior for on-policy distillation (`lmbda=1.0`) with **`gradient_accumulation_steps=2`**.
> 
> The test monkeypatches **`_reduce_divergence_loss`** to record the **`num_items_in_batch`** denominator used during **`train()`** and asserts it equals the **sum of completion tokens** across accumulation microbatches. It is expected to fail today because Hugging Face **`Trainer`** derives that count from raw dataloader labels (before generation replaces completions) and **`_RepeatBatchDataLoader`** repeats one batch per accumulation step, which **inflates the denominator** and **silently down-scales** loss and gradients.
> 
> **No trainer code changes** — documentation-only safety net until plan step 5.6 moves the count to GRPO-style **`_prepare_inputs`**; the test is written to assert on the applied denominator so it can be un-xfailed when that fix lands.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 23784c3a05429db535407257be033c40cb410d2b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/experimental/test_distillation_trainer.py [hunk 1]
```diff
@@ -622,6 +622,43 @@ def test_train_updates_params_on_and_off_policy(self, lmbda):
...
+            def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+                # Prompt positions are already -100, so this is the completion-token count either way.
+                recorded.append((num_items_in_batch, int((inputs["labels"] != -100).sum())))
+                return super().compute_loss(model, inputs, return_outputs, num_items_in_batch)
+
+        dataset = load_dataset("trl-internal-testing/zen", "conversational_language_modeling", split="train")
+        trainer = _RecordingTrainer(
+            model=self.model_id,
+            teacher_model=self.model_id,
+            args=self._make_args(gradient_accumulation_steps=2, max_steps=1),
```
Comment (cmpatino) [inline tests/experimental/test_distillation_trainer.py]: `_make_args` sets `lmbda=0.0`, so this test effectively tests the off-policy case.

Just wanted to flag in case we want to change the `_make_args` default or add a separate test for the on-policy case.

**Review by cmpatino — COMMENTED** [review]
Left a comment to discuss whether we should add a test for the on-policy case.

File: tests/experimental/test_distillation_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py]: Good catch, reworked it. It now runs on-policy (lmbda=1.0) so generation actually runs, and it captures the denominator at _reduce_divergence_loss (where it's applied) rather than the compute_loss argument, so it stays valid and turns green once the count moves to the GRPO-style _prepare_inputs.

File: tests/experimental/test_distillation_trainer.py [hunk 2]
```diff
@@ -630,6 +630,43 @@ def test_train_updates_params_on_and_off_policy(self, lmbda):
...
+            processing_class=self.tokenizer,
+        )
+
+        trainer.train()
+
+        assert len(recorded) == 2, "expected one compute_loss call per accumulation step"
+        reported = recorded[0][0]
+        assert reported is not None, "transformers did not pass num_items_in_batch"
+        # The denominator should be the number of tokens summed over the accumulation window.
+        assert int(reported) == sum(trained_on for _, trained_on in recorded)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py]: Fixed. The test now captures num_items_in_batch at _reduce_divergence_loss instead of the compute_loss argument.

File: tests/experimental/test_distillation_trainer.py [hunk 3]
```diff
@@ -630,6 +630,43 @@ def test_train_updates_params_on_and_off_policy(self, lmbda):
...
+            def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+                # Prompt positions are already -100, so this is the completion-token count either way.
+                recorded.append((num_items_in_batch, int((inputs["labels"] != -100).sum())))
+                return super().compute_loss(model, inputs, return_outputs, num_items_in_batch)
+
+        dataset = load_dataset("trl-internal-testing/zen", "conversational_language_modeling", split="train")
+        trainer = _RecordingTrainer(
+            model=self.model_id,
+            teacher_model=self.model_id,
+            args=self._make_args(gradient_accumulation_steps=2, max_steps=1),
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py]: Fixed. The test now runs on-policy (lmbda=1.0), so generation replaces the completions and the dataset-label-vs-generated-token mismatch is exercised (denominator 38 vs 128 tokens actually trained on).

File: tests/experimental/test_distillation_trainer.py [hunk 4]
```diff
@@ -630,6 +630,46 @@ def test_train_updates_params_on_and_off_policy(self, lmbda):
         for name, param in previous_params.items():
             assert not torch.equal(param, trainer.model.get_parameter(name)), f"Parameter {name} has not changed."
 
+    @pytest.mark.xfail(
+        strict=True,
+        reason="On-policy, num_items_in_batch is computed by transformers from the raw dataloader labels before "
+        "generation replaces the completions, and _RepeatBatchDataLoader repeats one generation batch across the "
+        "accumulation window, so the denominator the loss divides by does not equal the completion tokens actually "
+        "trained on. Un-xfail when the count moves to the GRPO-style _prepare_inputs (plan 5.6).",
+    )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py:638]: Fixed, reverted to a non-strict xfail, matching the rest of the suite (all xfail markers in tests/ are non-strict). The reason string still points to un-xfailing at plan 5.6 when the denominator moves to _prepare_inputs.
